pyat/at/plot/standalone.py

```diff
@@ -172,6 +175,116 @@ def plot_sigma(sigma, axis: tuple[str, str] = ('x', 'xp'), scale: float = 1.0,
     ax.set_ylabel('{0}{1}'.format(ax2['label'], ax2['unit']))
     return line
 
+def plot_RF_bucket_hamiltonian(ring, ct_range=None, dp_range=None, num_points=400,
+                               num_levels=41, plot_separatrix=True, **kwargs):
+    r"""Plot the resulting longitudinal Hamiltonian of a ring (defining the RF bucket).
+    The Hamiltonian is calculated by summing all the cavities in the ring.
+    Harmonic cavities are supported by the function.
+
+    Parameters:
+        ring: Lattice description
+        ct_range (tuple): Forced lower and upper ct values for the plot.
+        Default to :math:`\pm 1.1 \times C / (2h)`
+        dp_range (tuple): Forced lower and upper dp values for the plot.
+        Default to twice the RF acceptance of the bucket.
+        num_points (int): Number of points for 1D grid (ct/dp)
+        Default to 400.
+        num_levels (int): Number of levels for contour plot. Odd number of
+        levels allow to center the colorbar around 0.
+        Default to 41.
+        plot_separatrix (bool): Flag to plot the separatrix contour
+        (:math:`\mathcal{H}=0`).
+
+    Returns:
+        CT:   (num_points,num_points) array: ct grid
+        DP:    (num_points,num_points) array: dp grid
+        hamiltonian:   (num_points,num_points) array: Hamiltonian values
+        along (CT,DP) grid
+    """
+    # (ct, dp) grid calculation (defined around the main RF bucket)
+    if ct_range is None:
+        ct = numpy.linspace(-1.1 * ring.circumference/ring.harmonic_number/2,
+                     1.1 * ring.circumference/ring.harmonic_number/2, num=num_points)
+    else:
+        ct = numpy.linspace(ct_range[0], ct_range[1], num=num_points)
+    if dp_range is None:
+        ring.disable_6d()
+        U0 = ring.energy_loss
+        overvoltage = ring.rf_voltage / U0
+        rfa = numpy.sqrt(2 * U0 / (numpy.pi * ring.mcf * ring.harmonic_number * ring.energy) * \
+                      (numpy.sqrt(overvoltage**2 - 1) - numpy.arccos(1 / overvoltage)))
+        ring.enable_6d()
+        dp = numpy.linspace(-2 * rfa, 2 * rfa, num=num_points)
+    else:
+        dp = numpy.linspace(dp_range[0], dp_range[1], num=num_points)
+    CT, DP = numpy.meshgrid(ct, dp)
+
+    # Momentum compaction/phase slip factors computed up to third order
+    ring.disable_6d()
+    alpha = get_mcf(ring, fit_order=3, n_step=10)
+    ring.enable_6d()
```

Comment:
Now that the code is simpler, I come to the main point: I would prefer that the function does not modify its input arguments. Here, if a user enters a lattice with id_6d False, he will find 6d turned on after calling the function, which is unexpected. It would be safer to use copy=True when enabling/disabling 6d (while trying to minimise the number of switches…)

Reply:
I did not know the could return a shallow copy of a ring using this argument, thanks for bringing this to my attention. I reshuffled some parts of my function to compute the momentum compaction factor at the beginning and only use once the disable_6d() method on a shallow copy of the original ring. I agree the function should not modify the ring argument.